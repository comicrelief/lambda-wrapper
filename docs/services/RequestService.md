# RequestService

Provides access to components of the HTTP request being handled.

## Usage

Since Lambda Wrapper v2, the `RequestService` instance is no longer passed as an argument to your wrapped handler, and must be obtained via `di`.

```ts
import lambdaWrapper, { RequestService } from '@comicrelief/lambda-wrapper';

export default lambdaWrapper.wrap(async (di) => {
  const request = di.get(RequestService);
  // get the 'name' request parameter, defaulting to 'world' if not set
  const name = request.get('name', 'world');
  return ResponseModel.generate({}, 200, `Hello, ${name}`);
});
```

### Headers

- `getAllHeaders` returns an object containing all HTTP headers, with all keys lowercase
- `getHeader` returns the value of a single HTTP header
- `getAuthorizationToken` extracts a Bearer token from the `Authorization` header

### Body

For requests that submit data in their body (POST, PATCH, PUT),

- `getAll` parses the body according to the `Content-Type` header
- `get` fetches a single value from the body

### URL parameters

For other request methods without a body (GET, HEAD, DELETE),

- `getAll` returns an object containing all query string parameters
- `get` fetches a single query string parameter

For all requests,

- `getPathParameter` fetches a path parameter value

### Client info

Some limited information about the client making the request is available.

- `getIp` returns the request's source IP address
- `getUserBrowserAndDevice` returns user agent details
