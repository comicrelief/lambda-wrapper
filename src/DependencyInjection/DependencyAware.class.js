/**
 * DependencyAwareClass Class
 */
export default class DependencyAwareClass {
  /**
   * DependencyAwareClass constructor
   * @param di DependencyInjection
   */
  constructor(di) {
    this.di = di;
  }

  /**
   * Get Dependency Injection Container
   * @return DependencyInjection
   */
  getContainer() {
    return this.di;
  }
}
