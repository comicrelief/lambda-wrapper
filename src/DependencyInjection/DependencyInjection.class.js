import { DEFINITIONS, DEPENDENCIES } from '../Config/Dependencies';

/**
 * DependencyInjection class
 */
export default class DependencyInjection {
  /**
   * DependencyInjection constructor
   * @param configuration
   * @param event
   * @param context
   */
  constructor(configuration, event, context) {
    this.event = event;
    this.context = context;

    this.dependencies = {};
    this.configuration = configuration;

    for (let x = 0; x <= 1; x += 1) {
      // Iterate over lapper dependencies and add to container
      Object.keys(DEFINITIONS).forEach((dependencyKey) => {
        this.dependencies[dependencyKey] = new DEPENDENCIES[dependencyKey](this);
      });

      // Iterate over child dependencies and add to container
      if (typeof configuration.DEPENDENCIES !== 'undefined') {
        Object.keys(configuration.DEPENDENCIES).forEach((dependencyKey) => {
          this.dependencies[dependencyKey] = new configuration.DEPENDENCIES[dependencyKey](this);
        });
      }
    }
  }

  /**
   * Get Dependency
   * @param definition
   * @return {*}
   */
  get(definition) {
    if (typeof this.dependencies[definition] === 'undefined') {
      throw Error(`${definition} does not exist in di container`);
    }

    return this.dependencies[definition];
  }

  /**
   * Get Event
   * @return {*}
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get Context
   * @return {*}
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Configuration
   * @param definition string
   * @return {*}
   */
  getConfiguration(definition = null) {
    if (definition !== null && typeof this.configuration[definition] === 'undefined') {
      return null;
    } else if (typeof this.configuration[definition] !== 'undefined') {
      return this.configuration[definition];
    }

    return this.configuration;
  }
}
