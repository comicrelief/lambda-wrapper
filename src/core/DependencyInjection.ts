import { Context } from 'aws-lambda';

import DependencyAwareClass from './DependencyAwareClass';
import { LambdaWrapperConfig } from './config';

// eslint-disable-next-line no-use-before-define
type Class<TInstance, TConfig extends LambdaWrapperConfig> = new (di: DependencyInjection<TConfig>) => TInstance;

/**
 * Dependency injection container.
 *
 * Dependencies (singleton instances of dependency-aware classes) are provided
 * to the main Lambda handler and other dependencies via this class.
 */
export default class DependencyInjection<TConfig extends LambdaWrapperConfig = any> {
  /**
   * Instantiated dependencies.
   */
  readonly dependencies: Record<string, DependencyAwareClass>;

  /**
   * True until all dependencies have been constructed.
   */
  private isConstructing = true;

  constructor(
    readonly config: TConfig,
    readonly event: any,
    readonly context: Context,
  ) {
    // get unique dependency classes -- a class may be included several times,
    // but should be instantiated only once
    const classes = Array.from(new Set(Object.values(config.dependencies)));

    // guard against duplicate keys
    const countByName = classes
      .map((Constructor) => Constructor.name)
      .reduce(
        (counts, name) => ({ ...counts, [name]: (counts[name] || 0) + 1 }),
        {} as Record<string, number>,
      );
    if (Object.values(countByName).some((count) => count > 1)) {
      const duplicateNames = Object.entries(countByName)
        .filter(([, count]) => count > 1)
        .map(([name]) => name);

      // if all class names are single-letter, they're probably minified -- in
      // this case, give a hint about how to fix it
      const action = duplicateNames.every((it) => it.length === 1)
        ? "If you don't recognise the single-letter names listed above, your "
          + "bundler may be minifying your code. You'll need to disable this "
          + 'for Lambda Wrapper to work correctly. Please refer to the Notes '
          + 'section of the Lambda Wrapper readme:\n\n'
          + '  https://github.com/comicrelief/lambda-wrapper#notes'
        : 'Please ensure that all dependency classes have a unique name.';

      throw new Error(
        `Dependency names are not unique: ${duplicateNames.join(', ')}\n\n${action}`,
      );
    }

    // instantiate all dependencies
    this.dependencies = Object.fromEntries(
      classes.map((Constructor) => [Constructor.name, new Constructor(this)]),
    );

    this.isConstructing = false;
  }

  /**
   * Get the singleton instance of the given dependency.
   *
   * @param dependency
   */
  get<T extends DependencyAwareClass>(dependency: Class<T, TConfig>): T {
    if (this.isConstructing) {
      throw new Error(
        'Dependencies are not available in dependency class constructors.\n\n'
        + 'To fix this, call `di.get` in the function where the dependency is '
        + 'used instead of inside your constructor.',
      );
    }

    const name = dependency.name;

    if (!this.dependencies[name]) {
      throw new Error(
        `${name} does not exist in dependency container\n\n`
        + `Make sure you've included ${name} in the 'dependencies' key of your `
        + 'Lambda Wrapper config.',
      );
    }

    return this.dependencies[name] as T;
  }

  /**
   * Get the event passed to AWS Lambda.
   *
   * @deprecated Use `di.event` instead.
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get the AWS Lambda context object.
   *
   * @deprecated Use `di.context` instead.
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Lambda Wrapper configuration.
   *
   * @deprecated Use `di.config` instead.
   */
  getConfiguration() {
    return this.config;
  }

  /**
   * True if the function is being executed in `serverless-offline`.
   *
   * We use the following checks for this:
   *
   * - if there is no function ARN, or the ARN includes 'offline'
   * - if `process.env.USE_SERVERLESS_OFFLINE` is set
   *
   * TODO: This is nothing to do with dependency injection and should be moved
   *   somewhere else! Any ideas?
   */
  get isOffline(): boolean {
    return !this.context.invokedFunctionArn
      || this.context.invokedFunctionArn.includes('offline')
      || !!process.env.USE_SERVERLESS_OFFLINE;
  }

  /**
   * Get the `service-stage-` prefix added by Serverless to deployed Lambda
   * function names. This is handy when you want to invoke other functions,
   * without having to hardcode the service name and stage.
   *
   * The returned prefix includes a trailing dash. To get the deployed name of
   * another Lambda function, concatenate its serverless function name (its key
   * in `serverless.yml`) onto the prefix:
   *
   * ```js
   * const serverlessFunctionName = 'MyFunction';
   * const deployedName = `${di.getLambdaPrefix()}${serverlessFunctionName}`;
   * ```
   *
   * This function relies on looking at the currently running Lambda function's
   * resource name. It will not work correctly if the Lambda function has been
   * given a custom resource name.
   */
  getLambdaPrefix(): string {
    const stage = process.env.STAGE;
    if (!stage) {
      /* eslint-disable no-template-curly-in-string */
      throw new Error(
        'STAGE is not set\n\n'
        + 'Please add to your Lambda environment:\n\n'
        + '    STAGE: ${sls:stage}\n',
      );
    }
    if (!this.context.functionName) {
      throw new Error('Lambda function name is unavailable in context');
    }
    const [service] = this.context.functionName.split(`-${stage}-`);
    return `${service}-${stage}-`;
  }
}
