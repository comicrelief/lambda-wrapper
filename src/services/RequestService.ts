import QueryString from 'querystring';

import { APIGatewayProxyEvent } from 'aws-lambda';
import useragent from 'useragent';
import validate from 'validate.js/validate';
import XML2JS from 'xml2js';

import DependencyAwareClass from '../core/DependencyAwareClass';
import ResponseModel from '../models/ResponseModel';
import LoggerService from './LoggerService';

export const REQUEST_TYPES = {
  DELETE: 'DELETE',
  GET: 'GET',
  HEAD: 'HEAD',
  OPTIONS: 'OPTIONS',
  PATCH: 'PATCH',
  POST: 'POST',
  PUT: 'PUT',
};

export const HTTP_METHODS_WITHOUT_PAYLOADS = [
  REQUEST_TYPES.DELETE,
  REQUEST_TYPES.GET,
  REQUEST_TYPES.HEAD,
  REQUEST_TYPES.OPTIONS,
];

export const HTTP_METHODS_WITH_PAYLOADS = [
  REQUEST_TYPES.PATCH,
  REQUEST_TYPES.POST,
  REQUEST_TYPES.PUT,
];

// Define action specific error types
export const ERROR_TYPES = {
  VALIDATION_ERROR: new ResponseModel({}, 400, 'required fields are missing'),
};

export type RequestFile = {
  type: string;
  filename: string;
  contentType: string;
  content: string | Buffer;
};

/**
 * Provides access to components of the HTTP request being handled.
 */
export default class RequestService extends DependencyAwareClass {
  /**
   * Get a parameter from the request.
   *
   * @param parameter
   * @param ifNull Value to return if the parameter is not set.
   * @param requestType
   */
  get(parameter: string, ifNull?: string | null, requestType?: string): string | string[] | null {
    const queryParameters = this.getAll(requestType);

    if (queryParameters === null) {
      return ifNull ?? null;
    }

    return queryParameters[parameter] ?? ifNull ?? null;
  }

  /**
   * Get all HTTP headers included in the request.
   *
   * Header names are converted to lowercase.
   *
   * @returns An object with a key for each header.
   */
  getAllHeaders(): Record<string, string | undefined> {
    const event = this.di.event as APIGatewayProxyEvent;
    if (!event.headers) {
      return {};
    }
    return Object.fromEntries(
      Object.entries(event.headers)
        .map(([key, value]) => [key.toLowerCase(), value]),
    );
  }

  /**
   * Get an HTTP header from the request.
   *
   * The header name is case-insensitive.
   *
   * @param name The name of the header.
   * @param [whenMissing] Value to return if the header is missing.
   *   (default: empty string)
   */
  getHeader(name: string, whenMissing = ''): string {
    const headers = this.getAllHeaders();
    const lowerName = name.toLowerCase();
    return headers[lowerName] ?? whenMissing;
  }

  /**
   * Get an authorization token from the `Authorization` header.
   */
  getAuthorizationToken(): string | null {
    const authorization = this.getHeader('Authorization');
    if (!authorization) {
      return null;
    }

    const tokenParts = authorization.split(' ');
    const tokenValue = tokenParts[1];

    if (!(tokenParts[0].toLowerCase() === 'bearer' && tokenValue)) {
      return null;
    }

    return tokenValue;
  }

  /**
   * Get a path parameter, or all path parameters if no `parameter` is given.
   *
   * @param parameter
   * @param ifNull Value to return if the parameter is not set.
   */
  getPathParameter(parameter?: string, ifNull = {}): any {
    const event = this.getContainer().getEvent() as APIGatewayProxyEvent;

    // If no parameter has been requested, return all path parameters
    if (!parameter && typeof event.pathParameters === 'object') {
      return event.pathParameters;
    }

    // If a specifc parameter has been requested, return the parameter if it exists
    if (
      parameter
      && typeof event.pathParameters === 'object'
      && event.pathParameters !== null
      && typeof event.pathParameters[parameter] !== 'undefined'
    ) {
      return event.pathParameters[parameter];
    }

    return ifNull;
  }

  /**
   * Get all request parameters
   *
   * @param requestType
   */
  getAll(requestType?: string): any {
    const event = this.getContainer().getEvent() as APIGatewayProxyEvent;

    if (
      HTTP_METHODS_WITHOUT_PAYLOADS.includes(event.httpMethod)
      || HTTP_METHODS_WITHOUT_PAYLOADS.includes(requestType || '')
    ) {
      // get simple parameters
      const params: Record<string, string | string[] | undefined> = {
        ...event.queryStringParameters,
      };
      // add array parameters as arrays
      if (event.multiValueQueryStringParameters !== null) {
        Object.keys(params)
          .filter((key) => key.endsWith('[]'))
          .forEach((key) => {
            params[key] = event.multiValueQueryStringParameters?.[key];
          });
      }
      return params;
    }

    if (
      HTTP_METHODS_WITH_PAYLOADS.includes(event.httpMethod)
      || HTTP_METHODS_WITH_PAYLOADS.includes(requestType || '')
    ) {
      const contentType = this.getHeader('Content-Type');
      let queryParameters = {};

      if (contentType.includes('application/x-www-form-urlencoded')) {
        queryParameters = QueryString.parse(event.body as string);
      }

      if (contentType.includes('application/json')) {
        try {
          queryParameters = JSON.parse(event.body as string);
        } catch {
          queryParameters = {};
        }
      }

      if (contentType.includes('text/xml')) {
        XML2JS.parseString(event.body as string, (error, result) => {
          queryParameters = error ? {} : result;
        });
      }

      if (contentType.includes('multipart/form-data')) {
        queryParameters = this.parseForm(true);
      }

      return typeof queryParameters !== 'undefined' ? queryParameters : {};
    }

    return null;
  }

  /**
   * Fetch the request IP address
   */
  getIp(): string | null {
    const event = this.getContainer().getEvent();

    if (
      typeof event.requestContext !== 'undefined'
      && typeof event.requestContext.identity !== 'undefined'
      && typeof event.requestContext.identity.sourceIp !== 'undefined'
    ) {
      return event.requestContext.identity.sourceIp;
    }

    return null;
  }

  /**
   * Get user agent details from the `User-Agent` header.
   */
  getUserBrowserAndDevice() {
    const userAgent = this.getHeader('user-agent');
    if (!userAgent) {
      return null;
    }

    try {
      const agent = useragent.parse(userAgent);
      const os = agent.os.toJSON();

      return {
        'browser-type': agent.family,
        'browser-version': agent.toVersion(),
        'device-type': agent.device.family,
        'operating-system': os.family,
        'operating-system-version': agent.os.toVersion(),
      };
    } catch {
      this.di.get(LoggerService).label('user-agent-parsing-failed');
      return null;
    }
  }

  /**
   * Test a request against validation constraints.
   *
   * See [validate.js](https://validatejs.org/) for how to write constraints.
   *
   * @param constraints
   */
  validateAgainstConstraints(constraints: object): Promise<void> {
    const logger = this.di.get(LoggerService);

    return new Promise((resolve, reject) => {
      const validation = validate(this.getAll(), constraints);

      if (typeof validation === 'undefined') {
        resolve();
      } else {
        logger.label('request-validation-failed');
        const validationErrorResponse = ERROR_TYPES.VALIDATION_ERROR;
        validationErrorResponse.setBodyVariable('validation_errors', validation);
        reject(validationErrorResponse);
      }
    });
  }

  /**
   * Fetch the request multipart form.
   *
   * @param useBuffer Whether to return file content as a `Buffer`.
   */
  parseForm(useBuffer: boolean) {
    // todo: rewrite this to use a dedicated package and add error handling
    /* eslint-disable @typescript-eslint/no-non-null-assertion */

    const event = this.getContainer().getEvent() as APIGatewayProxyEvent;
    const boundary = RequestService.getBoundary(event) as string;

    const body = event.isBase64Encoded
      ? Buffer.from(event.body as string, 'base64').toString('binary').trim()
      : event.body as string;

    const result: Record<string, RequestFile | string> = {};
    body.split(boundary).forEach((item) => {
      if (/filename=".+"/g.test(item)) {
        const name = item.match(/name=".+";/g)![0].slice(6, -2);
        result[name] = {
          type: 'file',
          filename: item.match(/filename=".+"/g)![0].slice(10, -1),
          contentType: item.match(/Content-Type:\s.+/g)![0].slice(14),
          content: useBuffer
            ? Buffer.from(item.slice(item.search(/Content-Type:\s.+/g) + item.match(/Content-Type:\s.+/g)![0].length + 4, -4), 'binary')
            : item.slice(item.search(/Content-Type:\s.+/g) + item.match(/Content-Type:\s.+/g)![0].length + 4, -4),
        };
      } else if (/name=".+"/g.test(item)) {
        result[item.match(/name=".+"/g)![0].slice(6, -1)] = item.slice(item.search(/name=".+"/g) + item.match(/name=".+"/g)![0].length + 4, -4);
      }
    });

    /* eslint-enable @typescript-eslint/no-non-null-assertion */

    return result;
  }

  /**
   * Fetch the request AWS event Records
   */
  getAWSRecords() {
    const event = this.getContainer().getEvent();
    const eventRecord = event.Records && event.Records[0];

    if (typeof event.Records !== 'undefined' && typeof event.Records[0] !== 'undefined' && typeof eventRecord.eventSource !== 'undefined') {
      return eventRecord;
    }
    return null;
  }

  /**
   * Gets a value independently from the case of the key.
   *
   * @param object
   * @param key
   */
  static getValueIgnoringKeyCase(object: Record<string, string | undefined>, key: string): string | undefined {
    const foundKey = Object.keys(object)
      .find((currentKey) => currentKey.toLocaleLowerCase() === key.toLowerCase());
    return foundKey && object[foundKey];
  }

  /**
   * Returns the content type
   * assoiated with the request
   *
   * @param event
   */
  static getBoundary(event: APIGatewayProxyEvent): string | undefined {
    return this.getValueIgnoringKeyCase(event.headers, 'Content-Type')?.split('=')?.[1];
  }
}
