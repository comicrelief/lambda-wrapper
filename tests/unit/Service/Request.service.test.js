/* eslint-disable sonarjs/no-duplicate-string */
import QueryString from 'querystring';

import CONFIGURATION from '../../../src/Config/Dependencies';
import DependencyInjection from '../../../src/DependencyInjection/DependencyInjection.class';
import RequestService, { HTTP_METHODS_WITHOUT_PAYLOADS, HTTP_METHODS_WITH_PAYLOADS } from '../../../src/Service/Request.service';

const getContext = require('../../mocks/aws/context.json');
const baseEvent = require('../../mocks/aws/event.json');

const getEvent = (overrides = {}) => JSON.parse(JSON.stringify(({
  ...baseEvent,
  ...overrides,
})));

describe('Service/RequestService', () => {
  afterEach(() => jest.resetAllMocks());

  HTTP_METHODS_WITHOUT_PAYLOADS.forEach((httpMethod) => {
    describe(`HTTP ${httpMethod}`, () => {
      describe('.getAll', () => {
        it('should return all get parameters as an array', () => {
          const event = getEvent({ httpMethod });
          event.queryStringParameters.test = 123;
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.getAll()).toEqual(event.queryStringParameters);
        });
      });

      describe('.get', () => {
        it('should fetch a query parameter from an AWS event', () => {
          const event = getEvent({ httpMethod });
          event.queryStringParameters.test = 123;
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('test')).toEqual(event.queryStringParameters.test);
        });

        it('should fetch a query parameter from an AWS event when the request type is set', () => {
          const event = getEvent({ httpMethod });
          event.queryStringParameters.test = 123;
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('test', null, httpMethod)).toEqual(event.queryStringParameters.test);
        });

        it(`should return null from a non existent ${httpMethod} parameter from an AWS event`, () => {
          const event = getEvent({ httpMethod });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('fake')).toEqual(null);
        });

        it(`should return null from a non existent ${httpMethod} parameter from an AWS event when the request type is set`, () => {
          const event = getEvent({ httpMethod });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('fake', null, httpMethod)).toEqual(null);
        });
      });

      describe('.validateAgainstConstraints', () => {
        const constraints = {
          giftaid: {
            numericality: true,
          },
        };

        beforeEach(() => {
          // Mute Winston
          // eslint-disable-next-line no-underscore-dangle
          jest.spyOn(console._stdout, 'write').mockImplementation(() => {});
        });

        it('should resolve if there are no validation errors', async () => {
          const event = getEvent({ httpMethod });
          event.queryStringParameters.giftaid = 123;
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          await expect(request.validateAgainstConstraints(constraints)).resolves.toEqual(undefined);
        });

        it('should return a response containing validation errors if the data provided is incorrect', async () => {
          const event = getEvent({ httpMethod });
          event.queryStringParameters.giftaid = 'abc';
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          await expect(request.validateAgainstConstraints(constraints)).rejects.toMatchSnapshot();
        });
      });

      describe('.getUserBrowserAndDevice', () => {
        it('should return null with `headers === undefined`', () => {
          const event = getEvent({ httpMethod, headers: undefined });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.getUserBrowserAndDevice()).toEqual(null);
        });

        it('should return null with `headers === null`', () => {
          const event = getEvent({ httpMethod, headers: null });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.getUserBrowserAndDevice()).toEqual(null);
        });

        it('should return a prettified user agent', () => {
          const event = getEvent({ httpMethod });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));
          expect(request.getUserBrowserAndDevice()).toEqual({
            'browser-type': 'Safari',
            'browser-version': '9.1.1',
            'device-type': 'Other',
            'operating-system': 'Mac OS X',
            'operating-system-version': '10.11.5',
          });
        });
      });
    });
  });

  HTTP_METHODS_WITH_PAYLOADS.forEach((httpMethod) => {
    const getPayloadEvent = (overrides = {}) => {
      const event = getEvent({ httpMethod });
      event.headers['Content-Type'] = 'application/x-www-form-urlencoded';
      event.body = 'grant_type=client_credentials&response_type=token&token_format=opaque';

      return { ...event, ...overrides };
    };

    const queryParameters = QueryString.parse(getPayloadEvent().body);

    describe(`HTTP ${httpMethod}`, () => {
      describe('.getAll', () => {
        it('should return all post parameters as an array', () => {
          const event = getPayloadEvent();
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.getAll()).toEqual(queryParameters);
        });
      });

      describe('.get', () => {
        it('should fetch a request body parameter from an AWS event', () => {
          const event = getPayloadEvent();
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('grant_type')).toEqual(queryParameters.grant_type);
        });

        it('should fetch a request body parameter from an AWS event when the request type is set', () => {
          const event = getPayloadEvent();
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('grant_type', null, httpMethod)).toEqual(queryParameters.grant_type);
        });

        it('should return null from a non existent request body parameter from an AWS event', () => {
          const event = getPayloadEvent();
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          expect(request.get('fake')).toEqual(null);
        });

        it('should return null from a non existent request body parameter from an AWS event when the request type is set', () => {
          const event = getPayloadEvent();
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext), getEvent);

          expect(request.get('fake', null, httpMethod)).toEqual(null);
        });
      });

      describe('.validateAgainstConstraints', () => {
        const constraints = {
          giftaid: {
            numericality: true,
          },
        };

        beforeEach(() => {
          // Mute Winston
          // eslint-disable-next-line no-underscore-dangle
          jest.spyOn(console._stdout, 'write').mockImplementation(() => {});
        });

        it('should resolve if there are no validation errors', async () => {
          const event = getPayloadEvent({ body: 'giftaid=123' });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          await expect(request.validateAgainstConstraints(constraints)).resolves.toEqual(undefined);
        });

        it('should return a response containing validation errors if the data provided is incorrect', async () => {
          const event = getPayloadEvent({ body: 'giftaid=abc' });
          const request = new RequestService(new DependencyInjection(CONFIGURATION, event, getContext));

          await expect(request.validateAgainstConstraints(constraints)).rejects.toMatchSnapshot();
        });
      });
    });
  });

  describe('getAllHeaders()', () => {
    const event = getEvent();
    const di = new DependencyInjection(CONFIGURATION, event, getContext);
    const request = new RequestService(di);

    it('should return all headers from the event', () => {
      expect(request.getAllHeaders()).toStrictEqual(event.headers);
    });
  });

  describe('getHeader()', () => {
    const event = getEvent();
    const di = new DependencyInjection(CONFIGURATION, event, getContext);
    const request = new RequestService(di);

    it('should return the specified header', () => {
      expect(request.getHeader('Accept')).toEqual(event.headers.Accept);
    });

    it("should return '' by default if header is missing", () => {
      expect(request.getHeader('Authorization')).toEqual('');
    });

    it('should return `whenMissing` if header is missing', () => {
      expect(request.getHeader('Authorization', 'none')).toEqual('none');
    });

    it('should not be case-sensitive', () => {
      expect(request.getHeader('accept')).toEqual(event.headers.Accept);
    });
  });
});
