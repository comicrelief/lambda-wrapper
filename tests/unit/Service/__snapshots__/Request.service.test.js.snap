// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service/RequestService HTTP DELETE .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP GET .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP HEAD .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP OPTIONS .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP PATCH .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP POST .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;

exports[`Service/RequestService HTTP PUT .validateAgainstConstraints should return a response containing validation errors if the data provided is incorrect 1`] = `
ResponseModel {
  "body": Object {
    "data": Object {},
    "message": "required fields are missing",
    "validation_errors": Object {
      "giftaid": Array [
        "Giftaid is not a number",
      ],
    },
  },
  "code": 400,
}
`;
